kube-runtime/src/wait.rs

```diff
@@ -0,0 +1,95 @@
+use futures::TryStreamExt;
+use kube::{Api, Resource};
+use serde::de::DeserializeOwned;
+use snafu::{futures::TryStreamExt as _, Snafu};
+use std::fmt::Debug;
+
+use crate::watcher::{self, watch_object};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("failed to probe for whether the condition is fulfilled yet: {}", source))]
+    ProbeFailed {
+        #[snafu(backtrace)]
+        source: watcher::Error,
+    },
+}
+
+/// Watch an object, and Wait for some condition `cond` to return `true`.
+///
+/// `cond` is passed `Some` if the object is found, otherwise `None`.
+///
+/// # Errors
+///
+/// Fails if the type is not known to the Kubernetes API, or if the [`Api`] does not have
+/// permission to `watch` and `list` it.
+///
+/// Does *not* fail if the object is not found.
+pub async fn await_condition<K>(
+    api: Api<K>,
+    name: &str,
+    mut cond: impl FnMut(Option<&K>) -> bool,
+) -> Result<(), Error>
+where
+    K: Clone + Debug + Send + DeserializeOwned + Resource + 'static,
+{
+    watch_object(api, name)
+        .context(ProbeFailed)
+        .try_take_while(|obj| {
+            let result = !cond(obj.as_ref());
+            async move { Ok(result) }
+        })
+        .try_for_each(|_| async { Ok(()) })
+        .await
+}
+
+pub mod conditions {
+    use kube::Resource;
+
+    pub fn is_deleted<K: Resource>(uid: &str) -> impl Fn(Option<&K>) -> bool + '_ {
+        move |obj: Option<&K>| {
+            obj.map_or(
+                // Object is not found, success!
+                true,
+                // Object is found, but a changed uid would mean that it was deleted and recreated
+                |obj| obj.meta().uid.as_deref() != Some(uid),
+            )
+        }
+    }
+}
+
+pub mod delete {
+    use kube::{api::DeleteParams, Api, Resource};
+    use serde::de::DeserializeOwned;
+    use snafu::{OptionExt, ResultExt, Snafu};
+    use std::fmt::Debug;
+
+    use super::{await_condition, conditions};
+
+    #[derive(Snafu, Debug)]
+    pub enum Error {
+        NoUid,
+        Delete { source: kube::Error },
+        Await { source: super::Error },
+    }
+
+    #[allow(clippy::module_name_repetitions, clippy::missing_errors_doc)]
+    pub async fn delete_and_finalize<K: Clone + Debug + Send + DeserializeOwned + Resource + 'static>(
```

Comment:
To be fair, there's not much for the controller to do in that case, it's not like you can retry a finalization.

I'm also not really sure we could really do better here than [`tokio::time::timeout`](https://docs.rs/tokio/1.11.0/tokio/time/fn.timeout.html), unless this is about forcing the user to at least make a choice about timeouts.

However, maybe this is more of a sign that we need some way to introspect what the `Controller`(/`applier`) is doing at runtime? Maybe something linked to https://github.com/tokio-rs/console?

Reply:
Yeah, maybe `delete_and_finalize` does not need it, and future visibility focus is more important.
If we do force a timeout __choice__ on users, I think it would probably make the most sense to do in `await_condition`.